aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -264,74 +327,41 @@ Tensor _embedding_bag_dense_backward_cpu(const Tensor &grad_, const Tensor &indi
-    if (o > 0) {
-      counts_uniq[o] += counts_uniq[o - 1];
-    }
-    o++;
-  }
-
-  if (mode == MODE_MEAN || mode == MODE_SUM) {
-    #pragma omp parallel for if (numel > 1000)
-      for (int64_t i = 0; i < (int64_t)counts_uniq.size(); i++) {
-        int64_t start = i == 0 ? 0 : counts_uniq[i - 1];
-        int64_t index = indices_data[start];
-        for (int64_t j = start; j < counts_uniq[i]; j++) {
-          int64_t source = offset2bag_data[j];
-          double scale = 1.0;
-          if (scale_grad_by_freq) {
-            scale /= counts[indices_data[i]];
-          }
-          if (mode == 1) { // MODE_MEAN
-            if (offsets_.size(0) == 1) {
-              auto bag_size = indices.size(0);
-              scale /= bag_size;
+  auto counts = compute_counts(num_weights, indices_data, numel);
+  auto next_unique_index_idx =
+      compute_counts_uniq(num_weights, indices_data, numel, counts);
+
+  #pragma omp parallel for if (numel > 1000)
+    for (int64_t i = 0; i < (int64_t)next_unique_index_idx.size(); i++) {
+      int64_t start = i == 0 ? 0 : next_unique_index_idx[i - 1];
+      int64_t index = indices_data[start];
+      for (int64_t j = start; j < next_unique_index_idx[i]; j++) {
+        int64_t source = offset2bag_data[j];
+        double scale = 1.0;
+        if (scale_grad_by_freq) {
+          scale /= counts[indices_data[i]];
+        }
+        if (mode == 1) { // MODE_MEAN
+          if (offsets_.size(0) == 1) {
+            auto bag_size = indices.size(0);
+            scale /= bag_size;
+          } else {
+            if (source == offsets_.size(0) - 1) {
+              scale /= indices.size(0) - offsets_data[offsets_.size(0) - 1];
             } else {
-              if (source == offsets_.size(0) - 1) {
-                scale /= indices.size(0) - offsets_data[offsets_.size(0) - 1];
-              } else {
-                scale /= offsets_data[source + 1] - offsets_data[source];
-              }
+              scale /= offsets_data[source + 1] - offsets_data[source];
             }
           }
-          int64_t ddim = grad.size(1);
-          if (grad.scalar_type() == kFloat) {
-            auto igwd = index_grad_weight.data<float>();
-            auto gd = grad.data<float>();
-            THBlas_axpy<float>(ddim, (float)scale, gd + ddim * source, 1,
-                        igwd + ddim * index, 1);
-          } else if (grad.scalar_type() == kDouble) {
-            auto igwd = index_grad_weight.data<double>();
-            auto gd = grad.data<double>();
-            THBlas_axpy<double>(ddim, (double)scale, gd + ddim * source, 1,
-                         igwd + ddim * index, 1);
-          }
         }
+        int64_t ddim = grad.size(1);
+        AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "embedding_bag_backward", [&] {
+          auto igwd = index_grad_weight.data<scalar_t>();
+          auto gd = grad.data<scalar_t>();
+          THBlas_axpy<scalar_t>(ddim, (scalar_t)scale, gd + ddim * source, 1,
```

Comment:
I wonder if this is still necessary. It was added to deal with overhead introduced by the corresponding ATen function. Generally, we, of course, want to get rid of legacy TH code. This could be another nix thing to fix.

Reply:
The reason why EmbeddingBag uses axpy, to my understanding, is that it was faster than the alternative. It might be that it doesn't work as well now, but we can figure that out in a future change.